Purchases/Purchasing/StoreKit2/ProductsFetcherSK1.swift

```diff
@@ -0,0 +1,127 @@
+    private let queue = DispatchQueue(label: "ProductsFetcherSK1")
+    private var productsByRequests: [SKRequest: Set<String>] = [:]
+    private var completionHandlers: [Set<String>: [(Set<SKProduct>) -> Void]] = [:]
+
+    init(productsRequestFactory: ProductsRequestFactory = ProductsRequestFactory()) {
+        self.productsRequestFactory = productsRequestFactory
+    }
+
+    func products(withIdentifiers identifiers: Set<String>,
+                  completion: @escaping (Set<SKProduct>) -> Void) {
+
+        queue.async { [self] in
+            let productsAlreadyCached = self.cachedProductsByIdentifier.filter { key, _ in identifiers.contains(key) }
+            if productsAlreadyCached.count == identifiers.count {
+                let productsAlreadyCachedSet = Set(productsAlreadyCached.values)
+                Logger.debug(String(format: Strings.offering.products_already_cached, identifiers))
+                completion(productsAlreadyCachedSet)
+                return
+            }
+
+            if let existingHandlers = self.completionHandlers[identifiers] {
+                Logger.debug(String(format: Strings.offering.found_existing_product_request, identifiers))
+                self.completionHandlers[identifiers] = existingHandlers + [completion]
+                return
+            }
+
+            Logger.debug(
+                String(format: Strings.offering.no_cached_requests_and_products_starting_skproduct_request, identifiers)
+            )
+            let request = self.productsRequestFactory.request(productIdentifiers: identifiers)
+            request.delegate = self
+            self.completionHandlers[identifiers] = [completion]
+            self.productsByRequests[request] = identifiers
+            request.start()
+        }
+    }
+}
+
+extension ProductsFetcherSK1: SKProductsRequestDelegate {
+
+    func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
+        queue.async { [self] in
+            Logger.rcSuccess(Strings.network.skproductsrequest_received_response)
+            guard let requestProducts = self.productsByRequests[request] else {
+                Logger.error("requested products not found for request: \(request)")
+                return
+            }
+            guard let completionBlocks = self.completionHandlers[requestProducts] else {
+                Logger.error("callback not found for failing request: \(request)")
+                return
+            }
+
+            self.completionHandlers.removeValue(forKey: requestProducts)
+            self.productsByRequests.removeValue(forKey: request)
+
+            self.cacheProducts(response.products)
+            for completion in completionBlocks {
+                completion(Set(response.products))
+            }
+        }
+    }
+
+    func requestDidFinish(_ request: SKRequest) {
+        Logger.rcSuccess(Strings.network.skproductsrequest_finished)
+        request.cancel()
```

Comment:
If the request finished, why the `.cancel()`?

Reply:
ha, right. I should have added a comment, but this goes back a while. 
there's an iOS bug which causes a warning in sandbox:
https://github.com/RevenueCat/purchases-ios/issues/391

I'll add a comment